causallib/time_varying/GFormula.py

```diff
@@ -0,0 +1,218 @@
+                                        If None, will predict from first observed event (t.min()).
+                  timeline_end (int): Common end time-step. If provided, will generate survival curves up to 'timeline_end'
+                                      for all patients.
+                                      If None, will predict up to last observed event (t.max()).
+
+              Returns:
+                  pd.DataFrame: with time-step index, subject IDs (X.index) as columns and ??
+              """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def apply_noise(self):
+        pass
+
+    def _prepare_data(self, X, a, t, y):
+        return pd.DataFrame()
+
+
+    @staticmethod
+    def _predict_trajectory(self, X, a, t) -> pd.DataFrame:
+        """ Predicts the trajectories for all covariates X and treatment a.
+
+            Args:
+                  X (pd.DataFrame): Baseline covariate matrix of size (num_subjects, num_features).
+                  a (pd.Series): Treatment assignment of size (num_subjects,).
+                  t (pd.Series): Followup durations, size (num_subjects,).
+
+            Returns:
+                 pd.DataFrame: with time-step index, subject IDs (X.index and a) columns and
+                  point values for each column as entries
+        """
+        pass
+
+
+class GFormula(GFormulaBase):
+    """
+        GFormula class that is based on Monte Carlo Simulation for creating noise. #TODO more on this
+    """
+    def __init__(self,
+                 outcome_model,
+                 treatment_model,
+                 covariate_models,
+                 refit_models=True
+                 ):
+        super().__init__(outcome_model, treatment_model,
+                         covariate_models, refit_models)
+
+    def fit(self,
+            X: pd.DataFrame,
+            a: pd.Series,
+            t: pd.Series,
+            y: Optional[Any] = None,
+            refit_models=True,
+            **kwargs
+            ):
+
+        if kwargs is None:
+            kwargs = {}
+
+        #TODO preprocess data to fit in the model
+        if refit_models:
+            self.treatment_model.fit(X, a, y, **kwargs)
+
+            for cov in self.covariate_models:
+                self.covariate_models[cov].fit(X, a, y, **kwargs)
+
+        self.outcome_model.fit(X, a, y, **kwargs)
+        return self
+
+
+    def estimate_individual_outcome(self, X: pd.DataFrame, a: pd.Series, t: pd.Series, y: Optional[Any] = None,
+                                    timeline_start: Optional[int] = None,
+                                    timeline_end: Optional[int] = None) -> pd.DataFrame:
+
+        min_time = timeline_start if timeline_start is not None else int(t.min())
+        max_time = timeline_end if timeline_end is not None else int(t.max())
+
+        contiguous_times = pd.Series(data=range(min_time, max_time + 1), name=t.name)  # contiguous time steps for inference
```

Comment:
I think it might be possible to opt for a design in which the `time` is defined by the treatment assignment. Which on inference time is a desired treatment trajectory.